vlib/v/tests/go_array_wait_test.v

```
@@ -0,0 +1,38 @@
+fn f(x f64) f64 {
+	y := x * x
+	return y
+}
+
+fn test_array_thread_f64_wait() {
+	mut r := []thread f64{cap: 10}
+	for i in  0 .. 10 {
+		r << go f(f64(i) + 0.5)
+	}
+	x := r.wait()
+	assert x == [0.25, 2.25, 6.25, 12.25, 20.25, 30.25, 42.25, 56.25, 72.25, 90.25]
```

Hm, how should this comparison work with approximate floats? I'm surprised it worked on the test platform (maybe compiling the binary on a different system and then deploying it on the test platform would increase chances that this comparison finally does what it should - namely it should **not** work and panic).

---

These are not "approximate" floats. As long as floats (and all intermediate results) are binary fractions (fractions whose denominators are (not too big) powers of 2) they represent exact results.
Only fractions having other prime factors in the denominator lead to periodic binary numbers that have to be truncated at some point.
I've carefully chosen the numbers so they have exact binary representations.

---

That holds only for IEEE 754, but not for C semantics of `float` and `double` (C semantics is vague to support several other floating point standards including those not being base-2). We should definitely stick with C semantics or add a runtime detection of the underlying float standard to be IEEE 754 otherwise `assert`.

---

Well, C11 (ISO/IEC 9899:2011) strongly suggests IEC 60559 (ANSI/IEEE 754) binary floating point arithmetic. And even if there might be other compilers/platforms around we can safely *define* that V's types `float32` and `float64` always follow this standard.
There are thoughts to support decimal floats (either IEEE-754-2008 decimals or even arbitrary precision ones). But these will definitely have other names (e.g. `dec32`, `dec64`or `decimal`).

---

>And even if there might be other compilers/platforms around we can safely `define` that V's types `float32` and `float64` always follow this standard.

Let's do it then:

1. add a comment to the line with `==` on the array stating it assumes IEEE 754 base-2 format with carefully chosen numbers
2. create a separate issue in https://github.com/vlang/v/issues tracker for enforcing IEEE 754 base-2 as the **only** supported format - this mainly includes writing a runtime multiplatform initialization code for FPU with **hard panic/assert** if IEEE 754 base-2 is not supported/available, then it also includes thorough testing - I'd strongly recommend Void Linux which easily available and is known to have first-class support for PowerPC (from the oldest Apple Macs up to the newest Power9 and Power10 architectures) and uses also musl instead of glibc/...

---

Done in #8928.

BTW: Void Linux does not support powerpc (but arm32/64-musl - which I am running on my Raspberry PIs). Maybe you confused this with Alpine Linux (but they seem to have phased out ppc32)? OpenBSD and NetBSD are still options.

I still have my G3 iBook (dual USB) I bought from my first self-earned money after graduation. When OpenBSD-6.9 is released I'll give it a try...
But I'm pretty sure that used IEEE754 floats, too...

---

>BTW: Void Linux does not support powerpc (but arm32/64-musl - which I am running on my Raspberry PIs).

It actually does - here some info from one of the core devs:

https://translate.google.com/translate?sl=auto&tl=en&u=https://www.abclinuxu.cz/blog/enlightenment/2019/8/void-linux-a-powerpc-power-architektura-cast-2

(but admittedly I don't know whether it's fully supported or still beta)

So your G3 should work too :wink:

And thanks for the PR https://github.com/vlang/v/pull/8928 - as I pointed out though there must be a **runtime check for IEEE 754** (compile time check is a good additional measure to "fail fast", but by far not enough).

---

Btw. the main repo is https://voidlinux-ppc.org/